sessiontxn/staleread/provider.go

```diff
@@ -83,19 +85,54 @@ func (p *StalenessTxnContextProvider) OnInitialize(ctx context.Context, tp sessi
 	}
 }
 
+// activateStaleTxn first commit old transaction if needed, and then prepare and activate a transaction
+// with the staleness snapshot ts. After that, it sets the relevant context variables.
 func (p *StalenessTxnContextProvider) activateStaleTxn() error {
-	if err := p.sctx.NewStaleTxnWithStartTS(p.ctx, p.ts); err != nil {
+	var err error
+	if err = sessiontxn.CheckBeforeNewTxn(p.ctx, p.sctx); err != nil {
 		return err
 	}
-	p.is = p.sctx.GetSessionVars().TxnCtx.InfoSchema.(infoschema.InfoSchema)
-	if err := p.sctx.GetSessionVars().SetSystemVar(variable.TiDBSnapshot, ""); err != nil {
+
+	txnScope := kv.GlobalTxnScope
+	if err = p.sctx.PrepareTSFuture(p.ctx, sessiontxn.ConstantFuture(p.ts), txnScope); err != nil {
 		return err
 	}
 
-	txnCtx := p.sctx.GetSessionVars().TxnCtx
-	txnCtx.IsStaleness = true
-	txnCtx.InfoSchema = p.is
-	return nil
+	txnFuture := p.sctx.GetPreparedTxnFuture()
+	if txnFuture == nil {
```

Comment:
ts is prepared in line 97. I think we can guarantee txnFuture is not nil here. Maybe this line is not necessary

Reply:
Yes, it's unnecessary.